### PR TITLE
Remove awaitBusy in favor of assertBusy (#28450)

### LINKED ISSUE #28450: awaitBusy returns a boolean!?!
So I've been using `awaitBusy` incorrectly for a while and it looks like I'm not alone. I'd sort of figured it worked like `assertBusy` and would throw an exception if the thing didn't happen in time. It doesn't. It returns a boolean.

I scanned the usages and see a good 15% of calls to `awaitBusy` don't assert that the result is true. I figure most of those are bugs waiting to happen.

Most of the calls to `awaitBusy` ultimately `assertTrue` on the result. Maybe someone just goes and cleans all these up? They are sneaky enough that something like the logger usage check might be useful here. Or actually throwing the exception. I dunno. Discuss?

### PR DESCRIPTION
**Pull request by liketic**
Currently the method ```awaitBusy``` returns a boolean value which indicates if the result is success. We can replace it with ```assertBusy```. 

Closes #28450

### CODE REVIEW

**Comment by cbuescher** [comment]
@elasticmachine test this please

**Comment by liketic** [comment]
Thanks @cbuescher , there is a test was failed because the reallocation is not start, I don't know how to trigger the reallocation to happen, may I ask you to tell me how to fix that? Thank you so much! 👍

**Comment by cbuescher** [comment]
@liketic Do you have details about the failing test? I don't see the previous CI run in this PR anymore. 
The class name and the method would be great. Also, if you can reproduce it, these should be an output in the logs beginning with "REPRODUCE WITH:", if you find that you can copy it here so anybody reviewing the PR can see what is failing exactly.

**Comment by javanna** [comment]
retest this please

File: server/src/test/java/org/elasticsearch/index/shard/IndexShardOperationPermitsTests.java
Comment (javanna) [inline server/src/test/java/org/elasticsearch/index/shard/IndexShardOperationPermitsTests.java]: could you add a message here please?

File: server/src/test/java/org/elasticsearch/indices/IndicesLifecycleListenerIT.java
Comment (javanna) [inline server/src/test/java/org/elasticsearch/indices/IndicesLifecycleListenerIT.java]: these 3 could all become assertions right instead of if then fail?

File: server/src/test/java/org/elasticsearch/indices/recovery/IndexRecoveryIT.java
Comment (javanna) [inline server/src/test/java/org/elasticsearch/indices/recovery/IndexRecoveryIT.java]: maybe split this in two lines and use `assertFalse` ?

File: server/src/test/java/org/elasticsearch/indices/store/IndicesStoreIntegrationIT.java
Comment (javanna) [inline server/src/test/java/org/elasticsearch/indices/store/IndicesStoreIntegrationIT.java]: I wonder if we could adapt this test so that we don't the empty catch blocks? If we don't always assert on the same result, maybe we can pass in a consumer that asserts on the result and we would not need to return the boolean anymore?

File: server/src/test/java/org/elasticsearch/persistent/TestPersistentTasksPlugin.java
Comment (javanna) [inline server/src/test/java/org/elasticsearch/persistent/TestPersistentTasksPlugin.java]: let's split this in multiple asserts?

File: server/src/test/java/org/elasticsearch/persistent/TestPersistentTasksPlugin.java
Comment (javanna) [inline server/src/test/java/org/elasticsearch/persistent/TestPersistentTasksPlugin.java]: could you explain why this is needed?

File: test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java
Comment (javanna) [inline test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java]: add messages here and above?

File: server/src/test/java/org/elasticsearch/action/admin/cluster/node/tasks/CancellableTasksTests.java
Comment (javanna) [inline server/src/test/java/org/elasticsearch/action/admin/cluster/node/tasks/CancellableTasksTests.java]: add a message?

**Comment by javanna** [comment]
retest this please

**Review by javanna — CHANGES_REQUESTED** [review]
hi @liketic thanks a lot for your PR, I left some comments, but it looks good already!

File: server/src/test/java/org/elasticsearch/indices/store/IndicesStoreIntegrationIT.java
Comment (liketic) [reply server/src/test/java/org/elasticsearch/indices/store/IndicesStoreIntegrationIT.java]: I have no idea about how to remove the catch blocks.  WDYT?

File: server/src/test/java/org/elasticsearch/persistent/TestPersistentTasksPlugin.java
Comment (liketic) [reply server/src/test/java/org/elasticsearch/persistent/TestPersistentTasksPlugin.java]: This is unnecessary. I have reverted it.

File: server/src/test/java/org/elasticsearch/persistent/TestPersistentTasksPlugin.java
Comment (liketic) [reply server/src/test/java/org/elasticsearch/persistent/TestPersistentTasksPlugin.java]: Seems like it's not possible to split to multiple asserts?

File: server/src/test/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreIT.java
Comment (liketic) [inline server/src/test/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreIT.java]: @javanna This assert is failed. I don't know how to make it passed. Could you help me?

**Comment by liketic** [comment]
Thanks @javanna . I pushed 86e25e80123fd6e7b7095423420853669b17c46e . For other comments, I replied under the comment.

File: server/src/test/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreIT.java
Comment (javanna) [reply server/src/test/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreIT.java]: silly one, but took me a while to figure this out... we were previously not asserting on the return value of `awaitBusy`, hence the wait was not really doing anything, only sleeping for ~300ms without any verifications. I think we can remove this wait from the test completely. It should be green without it.

**Review by javanna — CHANGES_REQUESTED** [review]
I left a few more comments, thanks a lot @liketic this is very close!

File: server/src/test/java/org/elasticsearch/indices/store/IndicesStoreIntegrationIT.java
Comment (liketic) [reply server/src/test/java/org/elasticsearch/indices/store/IndicesStoreIntegrationIT.java]: Thank you so much! It's much better now.

**Comment by liketic** [comment]
I pushed 4c97324712e9dd2af7110feefd4ad6d5a482cc89 .

**Comment by javanna** [comment]
retest this please

**Comment by javanna** [comment]
retest this please

**Comment by liketic** [comment]
@javanna I think the test should be green now..

**Comment by cbuescher** [comment]
This might need another nudge
@elasticmachine test this please

**Comment by liketic** [comment]
Thanks @cbuescher for picking up this. Does the test failure relate to my changes ?

**Comment by javanna** [comment]
heya @liketic sorry about the radio silence, I think that your PR is good to go, I would need to have a last look at it, but unfortunately we make heavy use of `awaitsBusy` in our x-pack code as well, hence we need to adapt that one too before we can merge this in. I started a while ago but then I got distracted and I didn't get it done yet. Will keep you posted. Thanks for your PR once again, it's great!

**Comment by javanna** [comment]
With the opening of x-pack, I took the chance to merge master in. This will make it clearer what tests need adapting. @liketic you can also give it a go if you feel like it, otherwise we will take it over.

**Comment by javanna** [comment]
test this please

**Comment by liketic** [comment]
@javanna Please review again!

**Comment by javanna** [comment]
that was quick @liketic . I gave this a try a while ago and it ended up being a timesync for me. We seem to be using `awaitsBusy` without ever checking the return value in multiple places, which pretty much means waiting for nothing. There were quite some subtleties in the tests that I looked at. I merged master in and running tests now.

**Comment by javanna** [comment]
test this please

**Comment by javanna** [comment]
retest this please

**Comment by liketic** [comment]
Hi @javanna , is that possible to trigger jenkins test again?

**Comment by tvernum** [comment]
@elasticmachine test this please

**Comment by cbuescher** [comment]
@liketic I saw this PR, it looks almost ready but had merge conflicts so I merged in master and fixed a few new places where "awaitsBusy" showed up. I hope you don't mind. 
@javanna probably needs a final review but I wanted to get the tests rolling to get this moving again.

**Comment by cbuescher** [comment]
@elasticmachine test this please

**Comment by liketic** [comment]
@cbuescher Can you trigger jenkins again?

**Comment by cbuescher** [comment]
@elasticmachine test this please

**Comment by tvernum** [comment]
@elasticmachine test this please

**Comment by tomcallahan** [comment]
Heya @liketic still interested in this one?  It seems the test failed.  If you update the PR, we're happy to rerun the tests

**Comment by tomcallahan** [comment]
@elasticmachine test this please

**Comment by tomcallahan** [comment]
looks like CI is passing now, @javanna how are you feeling about this one?

**Comment by javanna** [comment]
I agree that we should work on getting this PR merged. I recently had a chat with @danielmitterdorfer about this PR with the intention of handing it over to core/infra team. Back when I first reviewed this (a long while ago), I noticed that there were a lot of usages, especially in xpack, of `awaitBusy` that were not checking the returned `boolean` which is problematic. That was before xpack was opened and later @liketic was kind enough to also work on the xpack tests (thanks!!!).

I see mainly two types of these usages:

1. The following pretty much means `Thread.sleep(1000);` which is something we try not to do in our tests.

```
awaitBusy(() -> false, 1, TimeUnit.SECONDS);
```

The above is getting replaced in this PR with the following which does not improve things much:

```
        try {
            assertBusy(Assert::fail, 1, TimeUnit.SECONDS);
        } catch (AssertionError ignore) {
        }
```

I would much rather prefer to fix these tests to have some condition to wait for. I realize that it may take longer and maybe it could be done as a follow-up. I guess I just see less value in removing `awaitBusy` if it gets replaced with `assertBusy` used this way.

2. The following does wait for some condition, but we don't check what gets returned, which means that the test will happily go on despite the condition we waited for is not met:

```
        BooleanSupplier hasUpgradeCompleted = checkCompletionStatus(upgrade);
        awaitBusy(hasUpgradeCompleted);
```

The above is currently getting replaced in this PR with the following which looks better, yet given that we are now checking that the condition is met, the test may end up failing:

```
        CheckedRunnable<Exception> hasUpgradeCompleted = checkCompletionStatus(upgrade);
        assertBusy(hasUpgradeCompleted);
```

A green build should at least indicate that we are heading in the right direction, but with these tests that have waiting conditions, we should probably run the tests more times to be confident that all updated tests are working fine.

Maybe a good way forward would be to start a feature branch based on this PR, improve the tests further (addressing the concerns I raised above) which would also allow us to have more CI runs on it so we merge it when we are confident enough?

**Pull request closed without merging**
